File: python/Produk.py
# Saya Dhiya Ulhaq dengan NIM 2407716 mengerjakan Tugas Praktikum 
# (Konsep OOP & Enkapsulasi) dalam Bahasa Pemrograman "Python"
# dalam mata kuliah desain & pemrograman berorientasi objek untuk keberkahan-Nya 
# maka saya tidak akan melakukan kecurangan seperti yang telah di spesifikasikan

class Produk:
    # Class variable untuk menyimpan semua produk
    daftar_produk = []

    def __init__(self, id, nama, merek, model, harga, stok):
        # Konstruktor untuk membuat objek Produk baru
        self.id = id          # Menyimpan ID produk
        self.nama = nama      # Menyimpan nama produk
        self.merek = merek    # Menyimpan merek produk
        self.model = model    # Menyimpan model produk
        self.harga = harga    # Menyimpan harga produk
        self.stok = stok      # Menyimpan jumlah stok produk

    # Fungsi format harga
    @staticmethod
    def format_harga(harga):
        # Format ke ribuan dengan titik, desimal koma
        return f"{harga:,.2f}".replace(",", "X").replace(".", ",").replace("X", ".")

    # Getter
    def get_id(self):
        return self.id        # Mengembalikan ID produk

    def get_nama(self):
        return self.nama      # Mengembalikan nama produk

    def get_merek(self):
        return self.merek     # Mengembalikan merek produk

    def get_model(self):
        return self.model     # Mengembalikan model produk

    def get_harga(self):
        return self.harga     # Mengembalikan harga produk

    def get_stok(self):
        return self.stok      # Mengembalikan stok produk

    # Setter
    def set_nama(self, nama):
        self.nama = nama      # Mengubah nama produk

    def set_merek(self, merek):
        self.merek = merek    # Mengubah merek produk

    def set_model(self, model):
        self.model = model    # Mengubah model produk

    def set_harga(self, harga):
        self.harga = harga    # Mengubah harga produk

    def set_stok(self, stok):
        self.stok = stok      # Mengubah stok produk

    # Tambah Data
    @classmethod # method bisa mengelola data yang bersifat global pada kelas (daftar_produk)
    def tambah_data(cls):
        # Menerima input dari user untuk membuat produk baru
        id = int(input("Masukkan ID: "))
        nama = input("Masukkan Nama Produk: ")
        merek = input("Masukkan Merek: ")
        model = input("Masukkan Model: ")
        harga = float(input("Masukkan Harga: "))
        stok = int(input("Masukkan Stok: "))

        # Membuat objek Produk baru dan menambahkannya ke daftar_produk
        cls.daftar_produk.append(Produk(id, nama, merek, model, harga, stok))
        print("Produk berhasil ditambahkan!")  # Konfirmasi penambahan

    # Cari Produk
    @classmethod
    def cari_produk(cls, id):
        # Mencari produk berdasarkan ID
        for i, p in enumerate(cls.daftar_produk):
            if p.get_id() == id:
                return i  # Mengembalikan index produk jika ditemukan
        return -1        # Mengembalikan -1 jika tidak ditemukan

    # Tampilkan Data
    @classmethod
    def tampilkan_data(cls):
        if not cls.daftar_produk:
            print("Belum ada data produk.")  # Pesan jika daftar kosong
            return

        header = ["ID", "Nama", "Merek", "Model", "Harga", "Stok"]

        # Menentukan lebar kolom berdasarkan header dan panjang data
        col_widths = [len(h) for h in header]
        for p in cls.daftar_produk:
            col_widths[0] = max(col_widths[0], len(str(p.get_id()))) # iD
            col_widths[1] = max(col_widths[1], len(p.get_nama())) # nama
            col_widths[2] = max(col_widths[2], len(p.get_merek())) # merek
            col_widths[3] = max(col_widths[3], len(p.get_model())) # model
            col_widths[4] = max(col_widths[4], len(Produk.format_harga(p.harga))) # harga
            col_widths[5] = max(col_widths[5], len(str(p.get_stok()))) # stok

        # Fungsi internal untuk mencetak garis tabel
        def garis():
            print("+" + "+".join(["-" * (w + 2) for w in col_widths]) + "+")

        # Menampilkan header tabel
        print("\n=== DAFTAR PRODUK ELEKTRONIX ===")
        garis()
        print("| " + " | ".join(h.ljust(col_widths[i]) for i, h in enumerate(header)) + " |")
        garis()

        # Menampilkan semua data produk
        for p in cls.daftar_produk:
            row = [
                str(p.get_id()).ljust(col_widths[0]),
                p.get_nama().ljust(col_widths[1]),
                p.get_merek().ljust(col_widths[2]),
                p.get_model().ljust(col_widths[3]),
                Produk.format_harga(p.harga).ljust(col_widths[4]),
                str(p.get_stok()).ljust(col_widths[5]),
            ]
            print("| " + " | ".join(row) + " |")
        garis()  # Garis penutup tabel

    # Cari Data
    @classmethod
    def cari_data(cls):
        id = int(input("Masukkan ID Produk yang ingin dicari: "))
        index = cls.cari_produk(id)  # Memanggil fungsi cari_produk
        if index == -1:
            print("Produk tidak ditemukan!")  # Produk tidak ada
        else:
            p = cls.daftar_produk[index]  # Produk ditemukan

            # menampilkan data
            header = ["ID", "Nama", "Merek", "Model", "Harga", "Stok"]
            col_widths = [
                max(len(header[0]), len(str(p.get_id()))),
                max(len(header[1]), len(p.get_nama())),
                max(len(header[2]), len(p.get_merek())),
                max(len(header[3]), len(p.get_model())),
                max(len(header[4]), len(Produk.format_harga(p.harga))),
                max(len(header[5]), len(str(p.get_stok()))),
            ]

            def garis():
                print("+" + "+".join(["-" * (w + 2) for w in col_widths]) + "+")

            # Menampilkan data produk yang ditemukan
            print("\nProduk ditemukan:")
            garis()
            print("| " + " | ".join(header[i].ljust(col_widths[i]) for i in range(len(header))) + " |")
            garis()
            row = [
                str(p.get_id()).ljust(col_widths[0]),
                p.get_nama().ljust(col_widths[1]),
                p.get_merek().ljust(col_widths[2]),
                p.get_model().ljust(col_widths[3]),
                Produk.format_harga(p.harga).ljust(col_widths[4]),
                str(p.get_stok()).ljust(col_widths[5]),
            ]
            print("| " + " | ".join(row) + " |")
            garis()

    # Update Data
    @classmethod
    def update_data(cls):
        id = int(input("Masukkan ID produk yang ingin diupdate: "))
        index = cls.cari_produk(id)
        if index == -1:
            print("Produk tidak ditemukan!")  # Produk tidak ada
            return

        # Input data baru
        nama = input("Masukkan Nama baru: ")
        merek = input("Masukkan Merek baru: ")
        model = input("Masukkan Model baru: ")
        harga = float(input("Masukkan Harga baru: "))
        stok = int(input("Masukkan Stok baru: "))

        # Update objek produk di daftar_produk
        p = cls.daftar_produk[index]
        p.set_nama(nama)
        p.set_merek(merek)
        p.set_model(model)
        p.set_harga(harga)
        p.set_stok(stok)

        print("Produk berhasil diupdate!")  # Konfirmasi update

    # Hapus Data
    @classmethod
    def hapus_data(cls):
        id = int(input("Masukkan ID produk yang ingin dihapus: "))
        index = cls.cari_produk(id)
        if index == -1:
            print("Produk tidak ditemukan!")  # Produk tidak ada
            return

        cls.daftar_produk.pop(index)  # Hapus produk dari daftar
        print("Produk berhasil dihapus!")  # Konfirmasi hapus

File: java/Produk.java
// Saya Dhiya Ulhaq dengan NIM 2407716 mengerjakan Tugas Praktikum 
// (Konsep OOP & Enkapsulasi) dalam Bahasa Pemrograman "Java"
// dalam mata kuliah desain & pemrograman berorientasi objek untuk keberkahan-Nya 
// maka saya tidak akan melakukan kecurangan seperti yang telah di spesifikasikan

import java.util.ArrayList;
import java.util.Scanner;
import java.text.NumberFormat;  
import java.util.Locale;

public class Produk {
    // Atribut private untuk menyimpan data produk
    private int id;       // ID unik produk
    private String nama;  // Nama produk
    private String merek; // Merek produk
    private String model; // Model produk
    private double harga; // Harga produk
    private int stok;     // Jumlah stok produk

    // ArrayList untuk menyimpan semua produk
    private final ArrayList<Produk> daftarProduk = new ArrayList<>();

    // Konstruktor default
    public Produk() {}

    // Konstruktor dengan parameter untuk membuat objek Produk baru
    public Produk(int id, String nama, String merek, String model, double harga, int stok) {
        this.id = id;
        this.nama = nama;
        this.merek = merek;
        this.model = model;
        this.harga = harga;
        this.stok = stok;
    }

    // Getter
    public int getId() { // Mengembalikan ID produk
        return id; 
    }        
    public String getNama() { // Mengembalikan nama produk
        return nama; 
    }
    public String getMerek() { // Mengembalikan merek produk
        return merek; 
    }
    public String getModel() { // Mengembalikan model produk
        return model; 
    }
    public double getHarga() { // Mengembalikan harga produk
        return harga; 
    }
    public int getStok() { // Mengembalikan stok produk
        return stok; 
    }

    // Setter
    public void setNama(String nama) { // Mengubah nama produk
        this.nama = nama; 
    }
    public void setMerek(String merek) { // Mengubah merek produk
        this.merek = merek; 
    }
    public void setModel(String model) { // Mengubah model produk
        this.model = model; 
    }
    public void setHarga(double harga) { // Mengubah harga produk
        this.harga = harga; 
    }
    public void setStok(int stok) {  // Mengubah stok produk
        this.stok = stok; 
    }

    // Tambah Data
    public void tambahData(Scanner sc) {
        // Meminta input user untuk setiap atribut produk
        System.out.print("Masukkan ID: ");
        this.id = sc.nextInt();
        System.out.print("Masukkan Nama Produk: ");
        this.nama = sc.next();
        System.out.print("Masukkan Merek: ");
        this.merek = sc.next();
        System.out.print("Masukkan Model: ");
        this.model = sc.next();
        System.out.print("Masukkan Harga: ");
        this.harga = sc.nextDouble();
        System.out.print("Masukkan Stok: ");
        this.stok = sc.nextInt();

        // Menambahkan produk baru ke daftarProduk
        daftarProduk.add(new Produk(id, nama, merek, model, harga, stok));

        // Menampilkan pesan sukses
        System.out.println("Produk berhasil ditambahkan!");
    }

    // Atribut formatter sebagai field class
    private final NumberFormat nf;

    {
        // initializer block → langsung dieksekusi saat objek dibuat
        nf = NumberFormat.getNumberInstance(new Locale("id", "ID"));
        nf.setMinimumFractionDigits(2);
        nf.setMaximumFractionDigits(2);
    }


    // Tampilkan Data
    public void tampilkanData() {
        // Jika daftarProduk kosong, tampilkan pesan dan hentikan
        if (daftarProduk.isEmpty()) {
            System.out.println("Belum ada data produk.");
            return;
        }

        // Header tabel
        String[] header = {"ID", "Nama", "Merek", "Model", "Harga", "Stok"};

        // Array untuk menyimpan lebar tiap kolom agar tabel rapi
        int[] colWidth = new int[header.length];
        for (int i = 0; i < header.length; i++) {
            colWidth[i] = header[i].length(); // Lebar minimal = panjang header
        }

        // Hitung panjang maksimal tiap kolom berdasarkan data produk
        for (Produk p : daftarProduk) {
            colWidth[0] = Math.max(colWidth[0], String.valueOf(p.getId()).length()); // ID
            colWidth[1] = Math.max(colWidth[1], p.getNama().length()); // nama
            colWidth[2] = Math.max(colWidth[2], p.getMerek().length()); // merek
            colWidth[3] = Math.max(colWidth[3], p.getModel().length()); // model
            colWidth[4] = Math.max(colWidth[4], nf.format(p.getHarga()).length()); // harga
            colWidth[5] = Math.max(colWidth[5], String.valueOf(p.getStok()).length()); // stok
        }

        // Membuat border tabel
        StringBuilder border = new StringBuilder("+");
        for (int w : colWidth) {
            border.append("-".repeat(w + 2)).append("+"); // Tambahkan padding + border
        }

        // Menampilkan header tabel
        System.out.println("\n=== DAFTAR PRODUK ELEKTRONIX ===");
        System.out.println(border);
        // cetak tabel
        System.out.print("|");
        for (int i = 0; i < header.length; i++) {
            System.out.printf(" %-" + colWidth[i] + "s |", header[i]);
        }
        System.out.println();
        System.out.println(border);

        // Menampilkan semua data produk
        for (Produk p : daftarProduk) {
            String hargaStr = nf.format(p.getHarga());
            System.out.printf("| %-" + colWidth[0] + "d ", p.getId());
            System.out.printf("| %-" + colWidth[1] + "s ", p.getNama());
            System.out.printf("| %-" + colWidth[2] + "s ", p.getMerek());
            System.out.printf("| %-" + colWidth[3] + "s ", p.getModel());
            System.out.printf("| %-" + colWidth[4] + "s ", hargaStr);
            System.out.printf("| %-" + colWidth[5] + "d |\n", p.getStok());
        }

        System.out.println(border); // Cetak border akhir
    }

    // Cari Produk
    private int cariProduk(int id) {
        // Loop untuk mencari index produk berdasarkan ID
        for (int i = 0; i < daftarProduk.size(); i++) {
            if (daftarProduk.get(i).getId() == id) {
                return i; // Mengembalikan index produk jika ditemukan
            }
        }
        return -1; // Jika tidak ditemukan, kembalikan -1
    }

    // cari data
    public void cariData(Scanner sc) {
        // input
        System.out.print("Masukkan ID Produk yang ingin dicari: ");
        int idCari = sc.nextInt();

        int index = cariProduk(idCari); // Cari index produk
        if (index == -1) {
            System.out.println("Produk tidak ditemukan!");
        } else {
            Produk p = daftarProduk.get(index); // Ambil produk yang ditemukan

            // Header dan lebar kolom seperti pada tampilkanData()
            String[] header = {"ID", "Nama", "Merek", "Model", "Harga", "Stok"};
            int[] colWidth = new int[header.length];
            for (int i = 0; i < header.length; i++) colWidth[i] = header[i].length();

            // Sesuaikan lebar kolom dengan data produk
            colWidth[0] = Math.max(colWidth[0], String.valueOf(p.getId()).length());
            colWidth[1] = Math.max(colWidth[1], p.getNama().length());
            colWidth[2] = Math.max(colWidth[2], p.getMerek().length());
            colWidth[3] = Math.max(colWidth[3], p.getModel().length());
            colWidth[4] = Math.max(colWidth[4], nf.format(p.getHarga()).length());
            colWidth[5] = Math.max(colWidth[5], String.valueOf(p.getStok()).length());

            // Buat border tabel
            StringBuilder border = new StringBuilder("+");
            for (int w : colWidth) border.append("-".repeat(w + 2)).append("+");

            System.out.println("\nProduk ditemukan:");
            System.out.println(border);

            // Cetak header
            System.out.print("|");
            for (int i = 0; i < header.length; i++) {
                System.out.printf(" %-" + colWidth[i] + "s |", header[i]);
            }
            System.out.println();
            System.out.println(border);

            // Cetak data produk
            String hargaStr = nf.format(p.getHarga());
            System.out.printf("| %-" + colWidth[0] + "d ", p.getId()); // ID
            System.out.printf("| %-" + colWidth[1] + "s ", p.getNama()); // nama
            System.out.printf("| %-" + colWidth[2] + "s ", p.getMerek()); // merek
            System.out.printf("| %-" + colWidth[3] + "s ", p.getModel()); // model
            System.out.printf("| %-" + colWidth[4] + "s ", hargaStr); // harga
            System.out.printf("| %-" + colWidth[5] + "d |\n", p.getStok()); // stok
            System.out.println(border);
        }
    }

    // Update Data
    public void updateData(Scanner sc) {
        System.out.print("Masukkan ID produk yang ingin diupdate: ");
        this.id = sc.nextInt(); // Ambil ID produk untuk diupdate

        int index = cariProduk(id); // Cari index produk
        if (index == -1) {
            System.out.println("Produk tidak ditemukan!");
            return;
        }

        // Ambil input baru dari user
        System.out.print("Masukkan Nama baru: ");
        this.nama = sc.next();
        System.out.print("Masukkan Merek baru: ");
        this.merek = sc.next();
        System.out.print("Masukkan Model baru: ");
        this.model = sc.next();
        System.out.print("Masukkan Harga baru: ");
        this.harga = sc.nextDouble();
        System.out.print("Masukkan Stok baru: ");
        this.stok = sc.nextInt();

        // Update data produk di daftarProduk
        Produk p = daftarProduk.get(index);
        p.setNama(nama);
        p.setMerek(merek);
        p.setModel(model);
        p.setHarga(harga);
        p.setStok(stok);

        System.out.println("Produk berhasil diupdate!");
    }

    // Hapus Data
    public void hapusData(Scanner sc) {
        System.out.print("Masukkan ID produk yang ingin dihapus: ");
        this.id = sc.nextInt(); // Ambil ID produk untuk dihapus

        int index = cariProduk(id); // Cari index produk
        if (index == -1) {
            System.out.println("Produk tidak ditemukan!");
            return;
        }

        daftarProduk.remove(index); // Hapus produk dari daftarProduk
        System.out.println("Produk berhasil dihapus!");
    }
}